### PR TITLE
Adding fmri example

### PR DESCRIPTION
**Pull request by ja-che**

### CODE REVIEW

**Comment by agramfort** [comment]
@ja-che can you look into the travis failures?

File: examples/plot_fmri_data_example.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    # Computing masked data
+    fmri_masked = masker.fit_transform(fmri_filename)
+    mask = masker.mask_img_.get_fdata().astype(bool)
+
+    X = np.asarray(fmri_masked)[condition_mask, :]
+    X = StandardScaler().fit_transform(X)
+
+    return Bunch(X=X, y=y, groups=groups, bg_img=bg_img, mask=mask,
+                 masker=masker)
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:79]: redundant to return both mask and masker ?

File: examples/plot_fmri_data_example.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+
+#############################################################################
+# Gathering and preprocessing Haxby dataset for a given subject
+# -------------------------------------------------------------
+list_subject = [1, 2, 3, 4, 5, 6]
+subject = list_subject[1]
+data = preprocess_haxby(subject=subject)
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:87]: lines 85-86 are useless

File: examples/plot_fmri_data_example.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+zscore_threshold_clust = zscore_from_pval((fwer_target / 2) * correction_clust)
+
+# Recovering the support with SVR decoder thresholded parametrically
+# We computed the regularization parameter by CV (C = 0.1)
+beta_hat, scale = standardized_svr(X, y, Cs=[0.1])
+pval, _, one_minus_pval, _ = pval_from_scale(beta_hat, scale)
+zscore_std_svr = zscore_from_pval(pval, one_minus_pval)
+
+# Recovering the support with SVR decoder thresholded by permutation test
+# This inference takes around 15 minutes.
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:123]: Oops. Isn't it possible to rely on another estimator for the sake of time ?

Comment (ja-che) [reply examples/plot_fmri_data_example.py:123]: You would like to a Ridge decoder thresholded by permutation test for example ?

Comment (bthirion) [reply examples/plot_fmri_data_example.py:123]: Yes, with a word to explain the reason: computation time.

File: examples/plot_fmri_data_example.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+#############################################################################
+# Function to fetch and preprocess Haxby dataset
+# ----------------------------------------------
+def preprocess_haxby(subject=2, memory=None, comparision='face-house'):
+    '''Gathering and preprocessing Haxby dataset for a given subject'''
+
+    # Gathering Data
+    haxby_dataset = datasets.fetch_haxby(subjects=[subject])
+    fmri_filename = haxby_dataset.func[0]
```
[the remark this answers is not in the document]
Comment (ja-che) [reply examples/plot_fmri_data_example.py:40]: No, this corresponds to all sessions. At line 41 - 44:

`    behavioral = pd.read_csv(haxby_dataset.session_target[0], sep=" ")

    conditions = pd.DataFrame.to_numpy(behavioral['labels'])
    session_label = pd.DataFrame.to_numpy(behavioral['chunks'])`

File: examples/plot_fmri_data_example.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    mask_img = haxby_dataset.mask
+    masker = NiftiMasker(mask_img=mask_img, standardize=True,
+                         smoothing_fwhm=None, memory=memory)
+
+    # Computing masked data
+    fmri_masked = masker.fit_transform(fmri_filename)
+    mask = masker.mask_img_.get_fdata().astype(bool)
+
+    X = np.asarray(fmri_masked)[condition_mask, :]
+    X = StandardScaler().fit_transform(X)
```
[the remark this answers is not in the document]
Comment (ja-che) [reply examples/plot_fmri_data_example.py:76]: Not really, I remove it.

File: examples/plot_fmri_data_example.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+from hidimstat.permutation_test import permutation_test_cv
+from hidimstat.gaonkar import gaonkar
+from hidimstat.clustered_inference import clustered_inference
+from hidimstat.ensemble_clustered_inference import ensemble_clustered_inference
+
+
+#############################################################################
+# Function to fetch and preprocess Haxby dataset
+# ----------------------------------------------
+def preprocess_haxby(subject=2, memory=None, comparision='face-house'):
```
[the remark this answers is not in the document]
Comment (ja-che) [reply examples/plot_fmri_data_example.py:35]: I added the possibility to select contrast in the example.

[the remark this answers is not in the document]
Comment (bthirion) [reply examples/plot_fmri_data_example.py]: Let users do that if they wish, but this is not your point here.

**Comment by agramfort** [comment]
brain figures are too dark and too small in https://67-259908992-gh.circle-artifacts.com/0/dev/auto_examples/plot_fmri_data_example.html#sphx-glr-auto-examples-plot-fmri-data-example-py

**Comment by ja-che** [comment]
@agramfort, @bthirion. I addressed your comments,  WDYT ?

**Comment by bthirion** [comment]
LGTM. Still a bit long to run...

**Comment by ja-che** [comment]
Should take a bit less than 2 minutes now !

**Comment by agramfort** [comment]
oui c'est très dense...

https://73-259908992-gh.circle-artifacts.com/0/dev/auto_examples/plot_fmri_data_example.html#sphx-glr-auto-examples-plot-fmri-data-example-py

yes it's really dense. Can you try to put the comments in section headers and split more the example. "Now we will do this. We can see ...." See it as an interactive paper @ja-che

**Comment by bthirion** [comment]
Besides "density" pointed by Alex, I'm fine with the example.

**Comment by ja-che** [comment]
Ok perfect, I will address the presentation problem. @agramfort do not do it please, I will do it this week end with high probability (>95%).

**Comment by ja-che** [comment]
@agramfort, @bthirion, I have increased the readability of the example, WDYT ? Thanks !

**Comment by agramfort** [comment]
Here is the rendered doc:

https://76-259908992-gh.circle-artifacts.com/0/dev/auto_examples/plot_fmri_data_example.html#sphx-glr-auto-examples-plot-fmri-data-example-py

I still feel it’s not balanced between code and pedagogical narrative.

Explain it like you would explain it to your grand parents :)

**Comment by bthirion** [comment]
Thx ! 
* To make the narration easier to follow, consider that readers do not understand "the uncompressed problem" you are talking about on line 4.
* more simply explain that you want to threshold classifier weights according to the confidence we have that they're nonzero
* take time to describe what you get when you run preprocess_haxby
* It is a bit weird to discuss the results in z-scale rather than p-val. Is there any motivation for that ? Oh yes, it's for display... I would put the p2z conversion in the figure generation loop.
* Explain what is seen at the end.
Assume that we use this script for teaching. What would we have to say ?
Best,

**Comment by ja-che** [comment]
Hello @bthirion and @agramfort,
I have added more comments and I think I have adressed all your requests. WDYT ? Thanks.

**Comment by ja-che** [comment]
Merged with last PR + all green !

File: examples/plot_fmri_data_example.py [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+that they are nonzero. Here, we work with the Haxby dataset and we focus on
+the 'face vs house' contrast. Thus, we consider the labeled activation maps
+of a given subject and try produce a brain map that corresponds to the
+discriminative pattern that makes the decoding of the two conditions.
+
+In this example, we show that standard statistical methods (i.e., method
+such as thresholding by permutation test the SVR or Ridge decoder or the
+algorithm proposed by Gaonkar [1]_) are not powerfull when applied on the
+uncompressed problem (i.e., the orignal problem in which the activation maps
+are not reduced using compression techniques such as parcelization).
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:17]:
```suggestion
are not reduced using compression techniques such as parcellation).
```

File: examples/plot_fmri_data_example.py [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+with statistical guarantees. More precisely, we aim at thresholding the
+weights of some estimated decoder maps according to the confidence we have
+that they are nonzero. Here, we work with the Haxby dataset and we focus on
+the 'face vs house' contrast. Thus, we consider the labeled activation maps
+of a given subject and try produce a brain map that corresponds to the
+discriminative pattern that makes the decoding of the two conditions.
+
+In this example, we show that standard statistical methods (i.e., method
+such as thresholding by permutation test the SVR or Ridge decoder or the
+algorithm proposed by Gaonkar [1]_) are not powerfull when applied on the
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:15]:
```suggestion
algorithm proposed by Gaonkar et al. [1]_) are not powerful when applied on the
```

File: examples/plot_fmri_data_example.py [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+uncompressed problem (i.e., the orignal problem in which the activation maps
+are not reduced using compression techniques such as parcelization).
+This is notably due to the high dimensionality (too many voxels) and
+structure of the data (too much correlation between neighboring voxels).
+We also present two methods that offer statistical guarantees but
+with a (small) spatial tolerance on the shape of the support:
+clustered desparsified lasso (CLuDL) combines clustering (parcelization)
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:22]:
```suggestion
clustered desparsified lasso (CLuDL) combines clustering (parcellation)
```

File: examples/plot_fmri_data_example.py [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+and statistical inference ; ensemble of clustered desparsified lasso (EnCluDL)
+adds a randomization step over the choice of clustering.
+
+EnCluDL is powerfull and does not depend on a unique clustering choice.
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:26]:
```suggestion
EnCluDL is powerful and does not depend on a unique clustering choice.
```

File: examples/plot_fmri_data_example.py [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+As shown in Chevalier et al. (2021) [2]_, for several task the estimated
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:27]:
```suggestion
As shown in Chevalier et al. (2021) [2]_, for several tasks the estimated
```

File: examples/plot_fmri_data_example.py [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+    fmri_masked = masker.fit_transform(fmri_filename)
+    X = np.asarray(fmri_masked)[condition_mask, :]
+
+    return Bunch(X=X, y=y, groups=groups, bg_img=bg_img, masker=masker)
+
+
+#############################################################################
+# Gathering and preprocessing Haxby dataset for a given subject
+# -------------------------------------------------------------
+# The `preprocess_haxby` fucntion make the preprocessing of the Haxby dataset,
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:108]:
```suggestion
# The `preprocess_haxby` function make the preprocessing of the Haxby dataset,
```

File: examples/plot_fmri_data_example.py [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+# computation time of the example.
+beta_hat, scale = standardized_svr(X, y, Cs=[0.1])
+pval_std_svr, _, one_minus_pval_std_svr, _ = pval_from_scale(beta_hat, scale)
+
+#############################################################################
+# Now, we compute p-values thanks to permutation tests applied to
+# 1/the weights of the SVR decoder or 2/the weights of the Ridge decoder.
+
+# To derive the p-values from the SVR decoder, you may change the next line by
+# `SVR_permutation_test_inference = True`. It should takes around 15 minutes.
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:149]:
```suggestion
# `SVR_permutation_test_inference = True`. It should take around 15 minutes.
```

File: examples/plot_fmri_data_example.py [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+# A faster solution is to compute the p-values by permutation test from the
+# Ridge decoder. The computation time is much shorter: around 20 seconds.
+
+estimator = Ridge()
+pval_corr_ridge_perm_test, one_minus_pval_corr_ridge_perm_test = \
+    permutation_test(X, y, estimator=estimator, n_permutations=200)
+
+#############################################################################
+# Now, let us run the Gaonkar algorithm (c.f. References).
+beta_hat, scale = gaonkar(X, y)
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:166]: Could we find a name different from `gaonkar` ? I think that this is a bit rude.

File: examples/plot_fmri_data_example.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+pval_gaonkar, _, one_minus_pval_gaonkar, _ = pval_from_scale(beta_hat, scale)
+
+#############################################################################
+# Now, the clustered inference algorithm which combines parcelization
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:170]:
```suggestion
# Now, the clustered inference algorithm which combines parcellation
```

File: examples/plot_fmri_data_example.py [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+# --------------------
+# To allow a better visualization of the disciminative pattern we will plot
+# z-maps rather than p-value maps. Assuming Gaussian distribution of the
+# estimators we can recover a z-score from a p-value by using the
+# inverse survival function.
+#
+# First, we set theoretical FWER target at 10%.
+
+n_samples, n_features = X.shape
+fwer_target = 0.1
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:198]: `target_fwer` sounds better

File: examples/plot_fmri_data_example.py [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+#############################################################################
+# Other methods need to be corrected. We consider the Bonferroni correction.
+# For methods that do not reduce the feature space, the correction
+# consists in dividing by the number of features.
+
+correction = 1. / n_features
+zscore_threshold_no_clust = zscore_from_pval((fwer_target / 2) * correction)
+
+#############################################################################
+# For methods that parcelizes the brain into groups of voxels, the correction
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:216]:
```suggestion
# For methods that parcelates the brain into groups of voxels, the correction
```

File: examples/plot_fmri_data_example.py [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+
+def plot_map(pval, one_minus_pval, zscore_threshold, title=None,
+             cut_coords=[-25, -40, -5], masker=masker, bg_img=data.bg_img):
+
+    zscore = zscore_from_pval(pval, one_minus_pval)
+    zscore_img = masker.inverse_transform(zscore)
+    plot_stat_map(zscore_img, threshold=zscore_threshold, bg_img=bg_img,
+                  dim=-1, cut_coords=cut_coords, title=title)
+
+    return
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:237]: This last line is useless

File: examples/plot_fmri_data_example.py [hunk 19]
```diff
@@ -0,0 +1,273 @@
...
+
+#############################################################################
+# Analysis of the results
+# -----------------------
+# As advocated in introduction, the methods that do not reduce the original
+# problem are not satisfying since they are too conservative.
+# Among those methods, the only one that makes discoveries is the one that
+# threshold the SVR decoder using a parametric approximation.
+# However this method has no statistical guarantees and we can see that some
+# isolated voxels are discovered and it seems quite unrealistic.
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:265]:
```suggestion
# isolated voxels are discovered, which seems quite spurious.
```

File: examples/plot_fmri_data_example.py [hunk 20]
```diff
@@ -0,0 +1,273 @@
...
+# The discriminative pattern derived from the clustered inference algorithm
+# (CluDL) show that the method is less conservative.
+# However, some unrealistic paterns are also included in this solution.
+# Finally, the solution provided by the ensemble clustered inference algorithm
+# (EnCluDL) seems realistic as we recover the visual cortex and do not make
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:270]:
```suggestion
# (EnCluDL) seems reasonable as we recover the visual cortex and do not make
```

File: examples/plot_fmri_data_example.py [hunk 21]
```diff
@@ -0,0 +1,273 @@
...
+# some dubious discoveries.
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:271]:
```suggestion
# spurious discoveries.
```

**Review by bthirion — COMMENTED** [review]
A set of minor issues, but looks good overall !

File: examples/plot_fmri_data_example.py [hunk 22]
```diff
@@ -0,0 +1,275 @@
...
+from sklearn.linear_model import Ridge
+from nilearn import datasets
+from nilearn.input_data import NiftiMasker
+from nilearn.image import mean_img
+from nilearn.plotting import plot_stat_map, show
+
+from hidimstat.stat_tools import zscore_from_pval, pval_from_scale
+from hidimstat.standardized_svr import standardized_svr
+from hidimstat.permutation_test import permutation_test, permutation_test_cv
+from hidimstat.gaonkar import ada_svr
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:60]: but you also need to change the module name  ;-)

File: examples/plot_fmri_data_example.py [hunk 23]
```diff
@@ -0,0 +1,275 @@
...
+
+if SVR_permutation_test_inference:
+    plot_map(pval_corr_svr_perm_test, one_minus_pval_corr_svr_perm_test,
+             zscore_threshold_corr, title='SVR permutation-test thresh.')
+
+plot_map(pval_corr_ridge_perm_test, one_minus_pval_corr_ridge_perm_test,
+         zscore_threshold_corr, title='Ridge permutation-test thresh.')
+
+plot_map(pval_ada_svr, one_minus_pval_ada_svr, zscore_threshold_no_clust,
+         title='Gaonkar algorithm')
```
Comment (bthirion) [inline examples/plot_fmri_data_example.py:253]: Maybe: ada-svr

**Review by bthirion — COMMENTED** [review]
Almost there !  thx,

**Comment by bthirion** [comment]
Thx ! This is great !
